test/fixtures/es-module-loaders/mock-loader.mjs

```diff
@@ -0,0 +1,137 @@
+import {receiveMessageOnPort} from 'worker_threads';
+const mockedModuleExports = new Map();
+let currentMockVersion = 0;
+
+/**
+ * FIXME: this is a hack to workaround loaders being
+ * single threaded for now
+ */
+function doDrainPort() {
+  let msg;
+  while (msg = receiveMessageOnPort(preloadPort)) {
+    onPreloadPortMessage(msg.message);
+  }
+}
+function onPreloadPortMessage({
+  mockVersion, resolved, exports
+}) {
+  currentMockVersion = mockVersion;
+  mockedModuleExports.set(resolved, exports);
+}
+let preloadPort;
+export function globalPreload({port}) {
+  preloadPort = port;
+  port.on('message', onPreloadPortMessage);
+  port.unref();
+  const insideAppContext = (getBuiltin, port, setImportMetaCallback) => {
+    let mockedModules = new Map();
+    let mockVersion = 0;
+    const doMock = (resolved, replacementProperties) => {
+      let exports = Object.keys(replacementProperties);
+      let namespace = Object.create(null);
+      let listeners = [];
+      for (const name of exports) {
+        let currentValue = replacementProperties[name];
+        Object.defineProperty(namespace, name, {
+          enumerable: true,
+          get() {
+            return currentValue;
+          },
+          set(v) {
+            currentValue = v;
+            for (let fn of listeners) {
+              try {
+                fn(name);
+              } catch {
+              }
+            }
+          }
+        });
+      }
+      mockedModules.set(resolved, {
+        namespace,
+        listeners
+      });
+      mockVersion++;
+      port.postMessage({ mockVersion, resolved, exports });
+      return namespace;
+    }
+    setImportMetaCallback((meta, context, parent) => {
+      if (context.url === 'node:mock') {
+        meta.doMock = doMock;
+        return;
+      }
+      if (context.url.startsWith('mock:')) {
+        let [proto, version, encodedTargetURL] = context.url.split(':');
+        let decodedTargetURL = decodeURIComponent(encodedTargetURL);
+        if (mockedModules.has(decodedTargetURL)) {
+          meta.mock = mockedModules.get(decodedTargetURL);
+          return;
+        }
+      }
+      parent(meta, context);
+    });
+  };
+  return `(${insideAppContext})(getBuiltin, port, setImportMetaCallback)`
+}
+
+
+// rewrites node: loading to mock: so that it can be intercepted
+export function resolve(specifier, context, defaultResolve) {
+  if (specifier === 'node:mock') {
+    return {
+      url: specifier
+    };
+  }
+  doDrainPort();
+  const def = defaultResolve(specifier, context);
+  if (context.parentURL?.startsWith('mock:')) {
+    // do nothing, let it get the "real" module
```

Comment:
The `mock:` prefix returns the _unmocked_ module?

Reply:
we could rename it to `unmocked:`? both are confusing to me

Reply:
Or `original:` maybe? Or `unmocked:` is fine. Jest calls it `actual`: https://jestjs.io/uk/docs/bypassing-module-mocks

Reply:
`actual:` seems fine

Reply:
re-reading, no `mock:` is the mocked version and without the `mock:` is the actual one @_o , this was done because there can be many versions of mocking the module so it needed to have things like a mock version number.

Reply:
😵‍💫

I thought the same as Geoffrey—I thought I left a question/comment about it, but I can't find it now.

I'm more confused by the re-affirmation. Why would a user import a mock in their source code?

Reply:
the user would never use `mock:` in a specifier, the underlying cache is using `mock:` for all the replacements of the "real" module. There can be N number of mocks of a given module, for example `mock:1:node%3Afs` would be 1 mocked version of `fs` while `mock:2:node%3Afs` would be a different mock, after the 1st one.

Basically this is needed for when multiple calls to `import.meta.mock('node:fs', ...)` create multiple modules. We can't actually replace the existing modules or unlink them from ESM graphs that already have been bound.

Reply:
Ohhhh, okay that makes sense! Might that be worth documenting?

I'm thinking this is too valuable to leave buried as a test fixture forever—at a later point, perhaps this should be surfaced in a more public-facing way (either in a doc, a quasi-official working example like our test loaders repo, etc).